File: Kosaraju_Algorithm.py
# Python implementation of Kosaraju's algorithm 
# It is used to print all Strongly Connected Components (SCCs) in a directed graph
# A component is said to be strongly connected if there is a path between all pairs of vertices in it
from collections import defaultdict 

# Adjacency list representation of directed graph using class - Graph 
class Graph: 
	def __init__(self,vertices): 
		self.V = vertices 
		self.graph = defaultdict(list) 

	# Function to add edge to graph 
	def addEdge(self,u,v): 
		self.graph[u].append(v) 
				
    # Function to fill vertices in stack 
	def fillStack(self,v,visited, stack): 
		# Mark current vertex as visited 
		visited[v] = True
		# Recur for vertices adjacent to this vertex 
		for i in self.graph[v]: 
			if visited[i] == False: 
				self.fillStack(i, visited, stack) 
		stack = stack.append(v) 
	
	# Function to return reverse of the graph 
	def getReverse(self): 
		g = Graph(self.V) 
		# Recur for all the vertices adjacent to this vertex 
		for i in self.graph: 
			for j in self.graph[i]: 
				g.addEdge(j,i) 
		return g 

    # function for DFS traversal
	def DFS(self,v,visited): 
		# Mark current vertex as visited and print it 
		visited[v] = True
		print(v, end = " ")
		# Recur for all vertices adjacent to this vertex 
		for i in self.graph[v]: 
			if visited[i] == False: 
				self.DFS(i,visited) 

	# Function to print all strongly connected components 
	def printSCC(self): 
		stack = [] 
		# Mark all the vertices as not visited (For first DFS) 
		visited = [False]*(self.V) 
		# Fill vertices in stack according to their finishing times 
		for i in range(self.V): 
			if visited[i] == False: 
				self.fillStack(i, visited, stack) 
		# Create a reversed graph 
		gr = self.getReverse() 
		# Mark all the vertices as not visited (For second DFS) 
		visited = [False]*(self.V) 
		# Process all vertices in order defined by Stack 
		while stack: 
			i = stack.pop() 
			if visited[i] == False: 
				gr.DFS(i, visited) 
				print() 


# Create Graph
v = int(input("Enter number of vertices in the graph: "))
g = Graph(v)
e = int(input("Enter number of edges in the graph: "))
for i in range(1,e+1):
    print("Edge no. %d - "%(i), end = "")
    num = list(map(int, input(' ').split()))
    g.addEdge(num[0],num[1])
print ("Strongly connected components in the graph are: ") 
g.printSCC() 
 

'''
Sample Input:
Enter number of vertices in the graph: 6
Enter number of edges in the graph: 7
Edge no. 1 - 0 1
Edge no. 2 - 1 2
Edge no. 3 - 2 0
Edge no. 4 - 2 3
Edge no. 5 - 3 4
Edge no. 6 - 4 5
Edge no. 7 - 5 3

Sample Output:
Strongly connected components in the graph are:
0 2 1
3 5 4
'''

File: Kosaraju_Algorithm/Kosaraju_algorithm.dart
/**
 *   Kosaraju's Algorithm
 * ----------------------------------------
 * 
 * This algorithm is used to find the number of strongly connected components
 * in a directed graph. It uses DFS. It performs DFS twice to achieve the goal based on the fact that
 * finishing time of starting vertex will be greater than the finishing times of vertices in other SCCs.
 * 
 */
// Importing required libraries
import 'dart:io';
import 'dart:collection';

// Class definition for a graph
class Graph{

    // Total number of vertices in a graph
    int numVertices;
    // Adjacency list of the graph
    var adj;

    // Constructor
    Graph(int v){
        this.numVertices = v;
        this.adj = List(this.numVertices);
        for(int i = 0; i < this.numVertices; i ++){
            adj[i] = List();
        }
    }

    // Method to add an edge to a graph
    void add_edge(int source, int destination){
        adj[source].add(destination);
    }

    // Utility method to fill the stack through DFS.
    void fillOrder(int v, List visited, Queue stack){
        visited[v] = true;
        var temp = this.adj[v];
        int m = temp.length;
        int i = 0;
        while(i < m){
            if(!visited[this.adj[v][i]]){
                fillOrder(this.adj[v][i], visited, stack);
            }
            i ++;
        }
        stack.add(v);
    }

    // Utility method to find the transpose of the graph
    Graph getTranspose(){

        Graph transpose = Graph(this.numVertices);

        for(int i = 0; i < transpose.numVertices; i ++){
            int destination = i;
            for(int j = 0; j < this.adj[i].length; j ++){
                transpose.adj[this.adj[i][j]].add(destination);
            }
        }

        return transpose;
    }

    // Utility method that performs DFS.
    void DFSUtil(int v, List visited, List result){
        visited[v] = true;
        result.add(v);
        int m = this.adj[v].length;

        for(int i = 0; i < m; i ++){
            if(!visited[this.adj[v][i]]){
                DFSUtil(this.adj[v][i], visited, result);
            }
        }
    }

    // Utility method to print strongly connected components.
    void printStronglyConnectedComponents(){
        // Using a LIFO queue as stack
        Queue stack = Queue();
        var visited = List(this.numVertices);

        // Initializing the visited list.
        for(int i = 0; i < this.numVertices; i ++){
            visited[i] = false;
        }

        // Filling the stack through DFS
        for(int i = 0; i < this.numVertices; i ++){
            if(visited[i] == false){
                fillOrder(i, visited, stack);
            }
        }

        // Finding transpose of the graph
        Graph g = getTranspose();
        for(int i = 0; i < this.numVertices; i ++){
            visited[i] = false;
        }

        
        int count = 0;

        // Finding strongly connected components using DFS on all vertices in stack.
        while(!stack.isEmpty){
            int v = stack.removeLast();

            if(visited[v] == false){
                var result = List();
                g.DFSUtil(v, visited, result);
                print("Strongly Connected Component ${count+1}:");
                print(result.join(" "));
                count += 1;
            }
        }
    }
}

// Driver method of the program
void main(){

    // Input of number of vertices
    print("Enter number of vertices in the graph:");
    var input = stdin.readLineSync();
    int n = int.parse(input);
    Graph g = Graph(n);

    // Input of number of edges
    print("Enter number of edges:");
    input = stdin.readLineSync();
    int m = int.parse(input);

    // Input of edges
    for(int i = 0; i < m; i ++){
        print("Edge ${i+1}:");
        print("Enter source vertex:");
        input = stdin.readLineSync();
        int source = int.parse(input);
        print("Enter destination vertex:");
        input = stdin.readLineSync();
        int destination = int.parse(input);
        g.add_edge(source, destination);
    }

    // Output of Strongly Connected Components.
    g.printStronglyConnectedComponents();

}

/*
 *    Sample Input and Output
 * -----------------------------
 * Enter number of vertices in the graph:
 * 8
 * Enter number of edges:
 * 9
 * Edge 1:
 * Enter source vertex:
 * 0
 * Enter destination vertex:
 * 1
 * Edge 2:
 * Enter source vertex:
 * 3
 * Enter destination vertex:
 * 0
 * Edge 3:
 * Enter source vertex:
 * 1
 * Enter destination vertex:
 * 2
 * Edge 4:
 * Enter source vertex:
 * 2
 * Enter destination vertex:
 * 3
 * Edge 5:
 * Enter source vertex:
 * 2
 * Enter destination vertex:
 * 4
 * Edge 6:
 * Enter source vertex:
 * 4
 * Enter destination vertex:
 * 5
 * Edge 7:
 * Enter source vertex:
 * 5
 * Enter destination vertex:
 * 6
 * Edge 8:
 * Enter source vertex:
 * 6
 * Enter destination vertex:
 * 7
 * Edge 9:
 * Enter source vertex:
 * 6
 * Enter destination vertex:
 * 4
 * Strongly Connected Component 1:
 * 0 3 2 1
 * Strongly Connected Component 1:
 * 4 6 5
 * Strongly Connected Component 1:
 * 7
 * 
 */
